base/gfspvgmgr/virtual_group_manager.go

```diff
@@ -451,3 +464,113 @@ func (vgm *virtualGroupManager) releaseSPAndGVGLoop() {
+
+	sps          map[uint32]*sptypes.StorageProvider // all sps,  id -> StorageProvider
+	unhealthySPs map[uint32]*sptypes.StorageProvider
+}
+
+func NewSPHealthChecker() *SPHealthChecker {
+	sps := make(map[uint32]*sptypes.StorageProvider)
+	unhealthySPs := make(map[uint32]*sptypes.StorageProvider)
+	return &SPHealthChecker{sps: sps, unhealthySPs: unhealthySPs}
+}
+
+func (checker *SPHealthChecker) AddSP(sp *sptypes.StorageProvider) {
+	checker.mutex.Lock()
+	defer checker.mutex.Unlock()
+
+	if _, exists := checker.sps[sp.GetId()]; exists {
+		// already exists
+	} else {
+		checker.sps[sp.GetId()] = sp
+	}
+}
+
+func (checker *SPHealthChecker) CheckSPHealth(sp *sptypes.StorageProvider) bool {
+	checker.mutex.Lock()
+	defer checker.mutex.Unlock()
+
+	if _, exists := checker.sps[sp.GetId()]; exists {
+		if _, unhealthyExists := checker.unhealthySPs[sp.GetId()]; unhealthyExists {
+			return false
+		} else {
+			return true
+		}
+	}
+	return false
+}
+
+func (checker *SPHealthChecker) checkALLSPHealth() {
+	spTemp := make(map[uint32]*sptypes.StorageProvider)
+	unhealthyTemp := make(map[uint32]*sptypes.StorageProvider)
+	checker.mutex.Lock()
+	for key, value := range checker.sps {
+		spTemp[key] = value
+	}
+	checker.mutex.Unlock()
+
+	unhealthyCnt := 0
+	for _, sp := range spTemp {
+		if !checker.isSPEndpointValid(sp.GetEndpoint()) {
+			unhealthyCnt++
+			unhealthyTemp[sp.GetId()] = sp
+		}
+	}
+
+	// more than 6 valid sp
+	if len(spTemp)-unhealthyCnt >= 6 {
+		checker.mutex.Lock()
+		checker.sps = spTemp
+		checker.unhealthySPs = unhealthyTemp
+		checker.mutex.Unlock()
+	}
+	log.Debugw("succeed to place these sp into unhealthy status", "unhealthy_sps", checker.unhealthySPs,
+		"sps", checker.sps, "unhealthy_sp", unhealthyCnt)
+}
+
+func (checker *SPHealthChecker) Start() {
+	go func() {
+		healthCheckerTicker := time.NewTicker(SPHealthCheckerInterval)
+		for range healthCheckerTicker.C {
+			log.Info("start to sp health checker")
+			checker.checkALLSPHealth()
+			log.Info("finish to sp health checker")
+		}
+	}()
+}
+
+func (checker *SPHealthChecker) isSPEndpointValid(endpoint string) bool {
+	ctxTimeout, cancel := context.WithTimeout(context.Background(), SPVerificationTimeout)
+	defer cancel()
+	client := &http.Client{
+		Transport: &http.Transport{
+			TLSClientConfig: &tls.Config{MinVersion: tls.VersionTLS12},
+		},
+		Timeout: SPVerificationTimeout * time.Second,
+	}
+
+	// Create an HTTP request to test the validity of the endpoint
+	urlToCheck := fmt.Sprintf("%s/status", endpoint)
+	req, err := http.NewRequestWithContext(ctxTimeout, "GET", urlToCheck, nil)
+	if err != nil {
+		return false
+	}
+
+	resp, err := client.Do(req)
+	if err != nil {
+		log.CtxErrorw(context.Background(), "failed to connect to sp", "endpoint", endpoint, "error", err)
+		return false
+	}
+	defer resp.Body.Close()
+
+	if resp.StatusCode != http.StatusOK {
+		log.CtxErrorw(context.Background(), "failed to check sp endpoint due to http status code:", resp.Status)
```

Comment:
fix as `log.CtxErrorw(context.Background(), "failed to check sp endpoint", "http status code", resp.Status)`.